### PR TITLE
Progress on event flow (animation frames)

### PR DESCRIPTION
**Pull request by raphlinus**
This is still a WIP because it's Windows-only, and needs some other refinement.

### CODE REVIEW

**Comment by raphlinus** [comment]
This is still something of a checkpoint because there's more work to be done on events, but it should be a self-contained chunk of improvements. (It's mac+Windows, and other rough edges have been refined a bit). I've also rebased it now that #83 landed.

**Comment by raphlinus** [comment]
Note: the CI failure is spurious, it's having trouble checking out rustfmt nightly. I think we might want to change our CI scripts so it only does stable rustfmt.

**Comment by zaynetro** [comment]
Could something like this help? https://docs.microsoft.com/en-ca/azure/devops/pipelines/test/flaky-test-management?view=azure-devops#flaky-test-management-and-reporting

I've also noticed that in the latest nightly components have been missing:

```
info: latest update on 2019-07-27, rust version 1.38.0-nightly (c43753f91 2019-07-26)
error: some components unavailable for download: 'rls', 'rustfmt' for channel 'nightly'
```

**Comment by raphlinus** [comment]
I wouldn't say the test is flaky, but in depending on nightly, and especially rustfmt nightly, we're sensitive to build failures. You can see those here: https://rust-lang-nursery.github.io/rust-toolstate/

I have some idea how to patch this, but would rather follow a "known good" recipe.

**Comment by zaynetro** [comment]
That's a useful page. Thanks!

File: src/event.rs [hunk 1]
```diff
@@ -28,6 +28,13 @@ pub enum Event {
     KeyUp(KeyEvent),
     Wheel(WheelEvent),
     HotChanged(bool),
```
[the remark this answers is not in the document]
Comment (raphlinus) [reply src/event.rs]: Good point, going forward I'll try to get these documented well, as they're an especially important part of the user-facing API. I've also been considering a stand-alone "event flow" document, but didn't get very much positive feedback and so am now leaning to extensive docstring. That's going to be more accessible anyway, I think.

File: src/lib.rs [hunk 2]
```diff
@@ -346,12 +390,19 @@ impl<T: Data> UiState<T> {
...
     }
 
-    /// Set the root widget as active, that is, receiving keyboard events.
+    /// Set the root widget as active.
+    ///
+    /// Warning: this is set as deprecated because it's not really meaningful.
+    /// It's likely that the intent was to set a default focus, but focus is
+    /// not yet implemented and there probably needs to be some other way to
+    /// identify the widget which should receive focus on startup.
+    #[deprecated]
```
[the remark this answers is not in the document]
Comment (raphlinus) [reply src/lib.rs]: Will do, but I think doesn't need to be in this PR.

**Comment by raphlinus** [comment]
Thanks!

Regarding the level of the api: this patch is *very explicitly* about the low level plumbing. I absolutely expect higher level abstractions to be built on top. The most direct (and likely most efficient) is for widgets to own animation objects, and plumb the `AnimFrame` event to them, along with the ctx so they can request anim frame if still animating. There are other structures, though. For relatively simple animating widgets, the animation can be a wrapper that encloses the widget.

Similarly, time is at a very low level here. Converting it into a higher level concept will be the responsibility of the animation object or wrapper. I've agonized here about the exact type for time; I think this is most useful for animations alone, but at the back of my head I'm also aware of the fact that it's not useful for, eg, synchronizing animations with audio. Time is *very complex,* because so much of the pipeline is async - events can happen with a timestamp and be queued, and frames painted go through a rendering, swapchain, and compositing pipeline before they're actually presented to the user. I'd like to come back to the question and do time in a very rigorous way, but I think that's a big project.

We should definitely have more discussion about what animation facilities are most useful now that the low level is in reasonably good shape (and whether they entail any refinements of the lower level).

**Pull request merged**